### ISSUE TITLE
VID API accept iterator instead of slice for payload

### ISSUE THREAD

**Issue by ggutoski** [EspressoSystems/jellyfish#417]
Needed in order to avoid unnecessary payload cloning in hotshot & sequencer. Should be easy---just restore old code.

**Comment by akonring** [comment]
> Needed in order to avoid unnecessary payload cloning in hotshot & sequencer. Should be easy---just restore old code.

tl;dr: I've tried to compile a list of discussion that relates to this issue below.
Think I could use some input on how/if to proceed. Or maybe I'm missing something; if this is as simple as stated above (@ggutoski , @jbearer).

***

We may want to change the API of VID so as to avoid collect/copy on the Hotshot side as described as described in this issue: https://github.com/EspressoSystems/HotShot/issues/2115.

Initial suggestion to deal with this issue is to change the VID API (`disperse` , `commit_only` ) to take the payload as an `Iterator<Item = u8>` . This suggestion is discussed here: https://github.com/EspressoSystems/HotShot/pull/2045#discussion_r1389985037(https://github.com/EspressoSystems/HotShot/pull/2045#discussion_r1389985037) but is less straight forward because Jellyfish VID needs the length of the payload not only to provide as returned metadata but also for doing proper chunking of the payload when exploiting the parallelism in `bytes_to_polys` [^1].

Another option to avoid copy of payload is to use [Cow](https://doc.rust-lang.org/std/borrow/enum.Cow.html) which is a smart pointer that clones the data lazily when mutation or ownership is required. This was discussed [here](https://github.com/EspressoSystems/hotshot-query-service/issues/266#issuecomment-1820937452) in the context of `transaction_with_proof`.

Discussions (https://github.com/EspressoSystems/espresso-sequencer/pull/1058#discussion_r1480442720) during sequencer payload redesign suggest lending just a reference when doing encode (`Encode: AsRef<[u8]>`). If possible, this seems like a bigger change across Espresso stack.

[^1]: chunking an `ExactSizeIterator` should indeed be possible but so far this has lead to a dead end - specifically because [ParallelBridge](https://docs.rs/rayon/latest/rayon/iter/trait.ParallelBridge.html) does not preserve the ordering of the elements.) Using `ExactSizeIterator` also poses another problem on how the `encode`-function knows the size without doing a linear scan over the payload.

**Comment by ggutoski** [comment]
I think `ExactSizeIterator` is not feasible. Will comment on the rest later.

**Comment by ggutoski** [comment]
https://github.com/EspressoSystems/HotShot/issues/2115
